### PR TITLE
implement mempool, work in progress

### PR DESCRIPTION
**Pull request by vlad9486**

### CODE REVIEW

**Review by tizoc — COMMENTED** [review]
@vlad9486 for writing to the socket directly to handle all that with mio the messages you are  interested in are defined in https://github.com/tezedge/tezedge/blob/develop/tezos/protocol-ipc-messages/src/lib.rs#L32-L35 (some parts imported from `tezos_api`, I think eventually those need to be moved, or the stuff here to `tezos/api` and both crates merged).

And then the message sent from the protocol runner to the shell/node process is defined in this same file as the `NodeMessage` enum.

In that case you will write and read directly from the socket without using the async api implemented in `tezos/protocol-ipc-client`.

One issue is that the handling of the subprocess (starting, cleanly shutting down, and restarting if something happens), is still handled outside, and if something happens the code inside the state machine will have no way to handle it. That is something that we have to think about how to handle (but it is a bit complicated right now because we are in a state of transition between the old and new architecture).

But for now I guess you will have to pass the path to the socket to the state machine so that it can make a new connection.

**Comment by tizoc** [comment]
Btw, here you can see how the messages are sent from the async client: https://github.com/tezedge/tezedge/blob/9ea5f403c4e60945f1c468e4909f688c9f2bbb6c/tezos/protocol-ipc-client/src/lib.rs#L627-L680

The macro handles a situation where the message received is not the expected one, but that is more about handling bugs than anything else, because right now that is not something that can happen after a timeout or some other situations like in the old IPC.

**Comment by tizoc** [comment]
Ah, one more detail, you will see that now there are two kind of messages, `*ForPrevalidation` and `*ForMempool`. The `*ForPrevalidation` messages are stateless, while the `*ForMempool` messages accumulate state. Depending on the situation you will want to use one kind or the other.

File: shell_automaton/src/service/protocol_service.rs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        self.responses.try_recv()
+            .map(|(response, id)| {
+                let handle = self.tasks.remove(id);
+                let _ = handle; // it is already done
+                response
+            })
+            .map_err(|_| ())
+    }
+
+    fn init_protocol_for_read(&mut self) {
```
[the remark this answers is not in the document]
Comment (vlad9486) [reply shell_automaton/src/service/protocol_service.rs:77]: @tizoc Sure. It was the simplest call to check that I really talking to protocol runner.

File: shell/src/chain_manager.rs [hunk 2]
```diff
@@ -781,59 +643,11 @@ impl ChainManager {
...
-                                    {
-                                        tell_peer(
-                                            &shell_automaton,
-                                            &peer.peer_id,
-                                            found_operation.as_ref().clone().into(),
-                                            &log,
-                                        );
-                                    }
-                                }
+                            PeerMessage::GetOperations(_) => {
```
Comment (tizoc) [inline shell/src/chain_manager.rs:646]: @vlad9486 there is no`TODO` here, but the body is empty.

**Pull request closed without merging**